### PR TITLE
[ARM32] Handle MovRelocatableImmediate as a special case (IF_T2_N3)

### PR DESCRIPTION
**Pull request by echesakovMSFT**
`IF_T2_N` instruction form now can be used in the three following cases:
1. *emitIns_R_I* mov => movw instruction conversion https://github.com/dotnet/coreclr/blob/47e106edc305c68651291d3862891ec6ad9f3ad6/src/jit/emitarm.cpp#L1845-L1851

2. *emitIns_R_I* for 16-bit immediates **or** relocatable immediates  https://github.com/dotnet/coreclr/blob/47e106edc305c68651291d3862891ec6ad9f3ad6/src/jit/emitarm.cpp#L1862-L1865

3. *emitIns_R_S* referencing a stack-based local variable
https://github.com/dotnet/coreclr/blob/47e106edc305c68651291d3862891ec6ad9f3ad6/src/jit/emitarm.cpp#L3516-L3519

The case 2 causes troubles for cross-bitness scenario (#16513) since we would have to use `ssize_t imm` almost everywhere in emitter and do explicit casting to `target_size_t` in many places as well. 

This PR introduces new `IF_T2_N3` instruction form which is used **only** for relocatable immediates. This also needs adding new instruction descriptor structure (`instrDescReloc` - the only one that can carry relocatable immediates in ARM32 emitter) and changing many places to be `target_ssize_t` instead of `ssize_t`

Running *jit-diff --pmi --altjit* and *jit-diff --crossgen x86_arm/crossgen.exe* shows no diffs for framework assemblies. 

Ubuntu arm pri1 and corefx tests have passed for these changes.

### CODE REVIEW

**Comment by echesakovMSFT** [comment]
@dotnet-bot help

**Comment by echesakovMSFT** [comment]
@dotnet-bot test Ubuntu arm Cross Checked normal Build and Test
@dotnet-bot test Ubuntu arm Cross Checked r2r Build and Test
@dotnet-bot test Ubuntu arm Cross Checked corefx_baseline Build and Test

@dotnet-bot test Windows_NT arm Cross Checked normal Build and Test
@dotnet-bot test Windows_NT arm Cross Checked r2r Build and Test
@dotnet-bot test Windows_NT arm Cross Checked corefx_baseline Build and Test

**Comment by echesakovMSFT** [comment]
**Windows_NT arm Cross Checked** legs failed due to #19020

**Comment by echesakovMSFT** [comment]
@dotnet/jit-contrib PTAL

**Comment by echesakovMSFT** [comment]
cc @dotnet/arm32-contrib

File: src/jit/emitarm.cpp [hunk 1]
```diff
@@ -5921,25 +5976,36 @@ size_t emitter::emitOutputInstr(insGroup* ig, instrDesc* id, BYTE** dp)
...
-                imm += (size_t)emitConsBlock;
-                if (!id->idIsCnsReloc() && !id->idIsDspReloc())
+                if (ins == INS_movw)
                 {
-                    goto SPLIT_IMM;
+                    imm &= 0xffff;
+                }
+                else
+                {
+                    imm = (imm >> 16) & 0xffff;
```
Comment (briansull) [inline src/jit/emitarm.cpp]: Add:
`assert(ins == INS_movt);`

File: src/jit/emitarm.cpp [hunk 2]
```diff
@@ -5948,26 +6014,34 @@ size_t emitter::emitOutputInstr(insGroup* ig, instrDesc* id, BYTE** dp)
                 {
                     imm = (imm >> 16) & 0xffff;
                 }
```
Comment (briansull) [inline src/jit/emitarm.cpp]: above Add:
`assert(ins == INS_movt);`

File: src/jit/emitarm.cpp [hunk 3]
```diff
@@ -5948,26 +6014,34 @@ size_t emitter::emitOutputInstr(insGroup* ig, instrDesc* id, BYTE** dp)
...
-            }
-
-            if (id->idIsCnsReloc() || id->idIsDspReloc())
-            {
-                assert((ins == INS_movt) || (ins == INS_movw));
+                code |= insEncodeImmT2_Mov(imm);
```
Comment (briansull) [inline src/jit/emitarm.cpp]: The encoding refactor is nice: `insEncodeImmT2_Mov`

File: src/jit/codegenarm.cpp [hunk 4]
```diff
@@ -78,7 +78,7 @@ void CodeGen::instGen_Set_Reg_To_Imm(emitAttr size, regNumber reg, ssize_t imm,
 
     if (EA_IS_RELOC(size))
     {
-        genMov32RelocatableImmediate(size, imm, reg);
+        genMov32RelocatableImmediate(size, (BYTE*)imm, reg);
```
Comment (BruceForstall) [inline src/jit/codegenarm.cpp]: Can you add a GitHub issue to track cleaning this up so we don't need to do this cast? I believe we determined that all the callers of instGen_Set_Reg_To_Imm that know it's relocatable already cast a pointer to this `ssize_t` imm.

File: src/jit/emitarm.cpp [hunk 5]
```diff
@@ -5921,53 +5976,74 @@ size_t emitter::emitOutputInstr(insGroup* ig, instrDesc* id, BYTE** dp)
...
-                    emitHandlePCRelativeMov32((void*)(dst - 8), (void*)imm);
             }
             else
             {
-                assert((imm & 0x0000ffff) == imm);
-                code |= (imm & 0x00ff);
-                code |= ((imm & 0x0700) << 4);
-                code |= ((imm & 0x0800) << 15);
-                code |= ((imm & 0xf000) << 4);
+                addr = emitConsBlock;
```
Comment (BruceForstall) [inline src/jit/emitarm.cpp]: In the N2 case, it seems like you could always calculate `addr = emitConsBlock + imm` and only in the "if" case do `imm == (size_t)addr`. If there are no relocs, then presumably you could also `assert(sizeof(size_t) == sizeof(target_size_t))`?

File: src/jit/emitarm.cpp [hunk 6]
```diff
@@ -6862,9 +6938,10 @@ void emitter::emitDispInsHelp(
         case IF_T2_L1:
         case IF_T2_L2:
         case IF_T2_N:
+        case IF_T2_N3:
             emitDispReg(id->idReg1(), attr, true);
             imm = emitGetInsSC(id);
-            if (fmt == IF_T2_N)
+            if (fmt == IF_T2_N || fmt == IF_T2_N3)
```
Comment (BruceForstall) [inline src/jit/emitarm.cpp]: Doesn't this need to call `emitGetInsRelocValue` for `IF_T2_N3`?

File: src/jit/codegenarm.cpp (hunk 4 above)
Comment (echesakovMSFT) [reply src/jit/codegenarm.cpp]: Thank for suggestion! Tracked as #19091

**Comment by echesakovMSFT** [comment]
@dotnet-bot test Ubuntu arm Cross Checked corefx_baseline Build and Test
@dotnet-bot test Ubuntu arm Cross Checked r2r Build and Test

**Comment by echesakovMSFT** [comment]
@dotnet-bot test Ubuntu arm Cross Checked normal Build and Test
@dotnet-bot test Ubuntu arm Cross Checked r2r Build and Test

@dotnet-bot test Windows_NT arm Cross Checked normal Build and Test
@dotnet-bot test Windows_NT arm Cross Checked r2r Build and Test

@dotnet-bot test Ubuntu arm Cross Release normal Build and Test
@dotnet-bot test Ubuntu arm Cross Release r2r Build and Test

@dotnet-bot test Windows_NT arm Cross Release normal Build and Test
@dotnet-bot test Windows_NT arm Cross Release r2r Build and Test

@dotnet-bot test Ubuntu arm Cross Checked corefx_baseline Build and Test
@dotnet-bot test Ubuntu arm Cross Checked corefx_jitstress1 Build and Test
@dotnet-bot test Ubuntu arm Cross Checked corefx_jitstress2 Build and Test
@dotnet-bot test Ubuntu arm Cross Checked corefx_minopts Build and Test

@dotnet-bot test Ubuntu arm Cross Checked jitstress1 Build and Test
@dotnet-bot test Ubuntu arm Cross Checked r2r_jitstress1 Build and Test
@dotnet-bot test Ubuntu arm Cross Checked jitstress2 Build and Test
@dotnet-bot test Ubuntu arm Cross Checked r2r_jitstress2 Build and Test
@dotnet-bot test Ubuntu arm Cross Checked minopts Build and Test

@dotnet-bot test Windows_NT arm Cross Checked corefx_baseline Build and Test
@dotnet-bot test Windows_NT arm Cross Checked corefx_jitstress1 Build and Test
@dotnet-bot test Windows_NT arm Cross Checked corefx_jitstress2 Build and Test
@dotnet-bot test Windows_NT arm Cross Checked corefx_minopts Build and Test

@dotnet-bot test Windows_NT arm Cross Checked jitstress1 Build and Test
@dotnet-bot test Windows_NT arm Cross Checked r2r_jitstress1 Build and Test
@dotnet-bot test Windows_NT arm Cross Checked jitstress2 Build and Test
@dotnet-bot test Windows_NT arm Cross Checked r2r_jitstress2 Build and Test
@dotnet-bot test Windows_NT arm Cross Checked minopts Build and Test

@dotnet-bot test Windows_NT x86_arm_altjit Checked corefx_baseline
@dotnet-bot test Windows_NT x86_arm_altjit Checked corefx_jitstress1
@dotnet-bot test Windows_NT x86_arm_altjit Checked corefx_jitstress2
@dotnet-bot test Windows_NT x86_arm_altjit Checked corefx_minopts

@dotnet-bot test Windows_NT x86_arm_altjit Checked jitstress1
@dotnet-bot test Windows_NT x86_arm_altjit Checked jitstress2
@dotnet-bot test Windows_NT x86_arm_altjit Checked minopts

@dotnet-bot test Windows_NT x86_arm_altjit Checked r2r_jitstress1
@dotnet-bot test Windows_NT x86_arm_altjit Checked r2r_jitstress2

**Comment by echesakovMSFT** [comment]
Re-trying failed legs (due to infrastructure)

@dotnet-bot test Ubuntu arm Cross Checked corefx_jitstress2 Build and Test
@dotnet-bot test Ubuntu arm Cross Checked corefx_minopts Build and Test 
@dotnet-bot test Ubuntu arm Cross Checked jitstress1 Build and Test 
@dotnet-bot test Ubuntu arm Cross Checked jitstress2 Build and Test
@dotnet-bot test Ubuntu arm Cross Checked minopts Build and Test 
@dotnet-bot test Ubuntu arm Cross Checked r2r Build and Test 
@dotnet-bot test Ubuntu arm Cross Checked r2r_jitstress2 Build and Test

**Comment by echesakovMSFT** [comment]
**Ubuntu arm Cross Checked corefx_baseline Build and Test** and **Ubuntu arm Cross Checked corefx_jitstress1 Build and Test** failed right after *All tests passed.* message due to infrastructure

**Comment by echesakovMSFT** [comment]
@dotnet-bot test Windows_NT x86_arm_altjit Checked r2r_jitstress1
@dotnet-bot test Windows_NT x86_arm_altjit Checked r2r_jitstress2

**Comment by echesakovMSFT** [comment]
@dotnet-bot test Ubuntu arm Cross Checked corefx_jitstress2 Build and Test
@dotnet-bot test Ubuntu arm Cross Checked corefx_minopts Build and Test

**Comment by echesakovMSFT** [comment]
@dotnet-bot test Ubuntu arm Cross Checked r2r_jitstress2 Build and Test

**Comment by echesakovMSFT** [comment]
@dotnet-bot test Ubuntu arm Cross Checked minopts Build and Test

**Comment by echesakovMSFT** [comment]
@dotnet-bot test Ubuntu arm Cross Checked jitstress1 Build and Test

**Comment by echesakovMSFT** [comment]
@dotnet-bot test Ubuntu arm Cross Checked r2r Build and Test

**Comment by echesakovMSFT** [comment]
Seems that all failures occur in master as well. I opened GitHub issues for new ones.

**Comment by echesakovMSFT** [comment]
@BruceForstall I addressed all your suggestions. PTAL

**Pull request merged**